common/xfer/plugins.go

```diff
@@ -0,0 +1,216 @@
+package xfer
+
+import (
+	"bytes"
+	"encoding/gob"
+	"fmt"
+	"sort"
+
+	"github.com/davecgh/go-spew/spew"
+	"github.com/mndrix/ps"
+	"github.com/ugorji/go/codec"
+
+	"github.com/weaveworks/scope/test/reflect"
+)
+
+type PluginSpec struct {
+	ID          string   `json:"id"`
+	Label       string   `json:"label"`
+	Description string   `json:"description,omitempty"`
+	Interfaces  []string `json:"interfaces"`
+}
+
+// PluginSpecs is a set of plugin specs keyed on ID. Clients must use
+// the Add method to add plugin specs
+type PluginSpecs struct {
+	psMap ps.Map
+}
+
+// EmptyPluginSpecs is the empty set of plugin specs.
+var EmptyPluginSpecs = PluginSpecs{ps.NewMap()}
+
+// MakePluginSpecs makes a new PluginSpecs with the given plugin specs.
+func MakePluginSpecs(specs ...PluginSpec) PluginSpecs {
+	return EmptyPluginSpecs.Add(specs...)
+}
+
+// Add adds the specs to the PluginSpecs. Add is the only valid way to grow a
+// PluginSpecs. Add returns the PluginSpecs to enable chaining.
+func (n PluginSpecs) Add(specs ...PluginSpec) PluginSpecs {
+	result := n.psMap
+	if result == nil {
+		result = ps.NewMap()
+	}
+	for _, spec := range specs {
+		result = result.Set(spec.ID, spec)
+	}
+	return PluginSpecs{result}
+}
+
+// Merge combines the two PluginSpecss and returns a new result.
+func (n PluginSpecs) Merge(other PluginSpecs) PluginSpecs {
+	nSize, otherSize := n.Size(), other.Size()
+	if nSize == 0 {
+		return other
+	}
+	if otherSize == 0 {
+		return n
+	}
+	result, iter := n.psMap, other.psMap
+	if nSize < otherSize {
+		result, iter = iter, result
+	}
```

Comment:
this datatype is a clone of `report.NodeSet`, so it got brought along with that. eugh.